### PR TITLE
feat(graindoc)!: Improve docgen for labeled & default arguments

### LINKED ISSUE #1763: GrainDoc: Labeled Arguments Get The Type `Option<a>` in param list
Labeled arguments get given the type `Option<a>` in the param list the generated graindoc as can be seen in the image below this should probably instead just be `a` given its defaulted to the same type or `?a` to represent that it is defaulted.

image
I think this miht have something todo with the todo that was left after the module update in the image below.
image

### PR DESCRIPTION
**Pull request by alex-snezhko**
Several improvements to graindoc w.r.t. labeled & default arguments.

- `@param` will now read the labels of parameters rather than their positions, meaning they can be written out of order as long as the names match. `@param unlabeled <argName>` can be used instead for unlabeled arguments, which uses the order of the parameter as was the old behavior (placeholder syntax for now, open to other ideas)
- Default parameters no longer show as `Option<a>` in graindoc output (closes #1763)
- Default parameters now have their default value shown in a new markdown column of the graindoc output if there are any default parameters for a function

### CODE REVIEW

File: compiler/graindoc/dune
Comment (phated) [inline compiler/graindoc/dune:5]:
```suggestion
 (libraries grain grain_diagnostics grain_utils))
```
When we print with printtyp or oprint.

File: compiler/src/diagnostics/graindoc_lexer.re
Comment (phated) [reply compiler/src/diagnostics/graindoc_lexer.re:56]: Alternatively, we could use `0`, `1`, `2`, etc for positional params

File: stdlib/array.gr
Comment (phated) [inline stdlib/array.gr:73]: Why'd you need to delete these? Should we instead change them to the labeled signature?

**Review by phated — CHANGES_REQUESTED** [review]
I'm SO excited for this! @alex-snezhko you have no idea how much I've dreamed about this graindoc implementation (probably since I created "optional arguments" issue). I love that we will look up docs by label and warn/error if they aren't correct ❤️ 

I have a few thoughts about the implementation but the big one is bringing in grain_formatting.

File: compiler/graindoc/docblock.re
Comment (alex-snezhko) [reply compiler/graindoc/docblock.re:278]: I used the printing code from the formatter as it seemed well-suited for the task of taking a `Parsetree.expression` and printing it out as a string with proper formatting

File: compiler/src/diagnostics/graindoc_lexer.re
Comment (alex-snezhko) [reply compiler/src/diagnostics/graindoc_lexer.re:56]: I think I like the idea of using numbers for positional params

File: compiler/src/parsing/parser.mly
Comment (alex-snezhko) [reply compiler/src/parsing/parser.mly:311]: Just because it seemed to conform better for what I was trying to do. This `expression` is only read when printing the default value of a parameter, and this value would always be filled in properly when parsing a function declaration. If I were to use an option here instead there would still more or less be an `Option.get` used to get the value

File: stdlib/array.gr
Comment (alex-snezhko) [reply stdlib/array.gr:73]: Related to #1775; I could change them to the labeled signature but I figured it would just be extraneous information since the param types and return type are already specified in the implementation

File: compiler/graindoc/docblock.re
Comment (phated) [reply compiler/graindoc/docblock.re:304]: That's the problem 🙃

**Comment by phated** [comment]
Blocked on the new formatter.

File: compiler/graindoc/docblock.re
Comment (alex-snezhko) [reply compiler/graindoc/docblock.re]: Update: formatter dep removed

**Comment by ospencer** [comment]
I think for now we should drop printing what the default value is—we chatted about this at the core team meeting and felt there were too many situations where the value provided wouldn't make sense, especially in the cases where the default references a binding or computes a value, where not having the surrounding context makes the default meaningless.

For now we can have a column that just reports a yes/no on if the value has a default, and maybe we can figure out something more clever in the future.

**Comment by alex-snezhko** [comment]
Should the column just be omitted altogether in favor of prefixing the parameter names with ? like they are in other places?

**Comment by ospencer** [comment]
Yeah, I think so.

File: compiler/graindoc/docblock.re
Comment (alex-snezhko) [reply compiler/graindoc/docblock.re]: Removed as no longer needed

**Comment by alex-snezhko** [comment]
Changed the format to include `?` as a prefix for parameters with default args rather than printing the default value

File: compiler/src/parsing/parsetree.re
Comment (phated) [inline compiler/src/parsing/parsetree.re:26]: did this get duplicated in some rebase or merge?

File: stdlib/pervasives.gr
Comment (phated) [inline stdlib/pervasives.gr:51]: Are these positional because we don't have labels on primitives? I believe @spotandjake is going to update all the primitives to have labels like he did in #1979

File: stdlib/runtime/dataStructures.gr
Comment (phated) [inline stdlib/runtime/dataStructures.gr:7]: Same question about primitives as previously

File: stdlib/runtime/wasi.gr
Comment (phated) [inline stdlib/runtime/wasi.gr:80]: Same question about primitives, does it also apply to foreigns?

File: compiler/src/typed/btype.re
Comment (phated) [inline compiler/src/typed/btype.re:411]: This should not be using a Parsetree type. We alias the Asttypes and then use the alias in the `typed/` directory

File: compiler/src/parsing/asttypes.re
Comment (phated) [inline compiler/src/parsing/asttypes.re:123]: Why was this moved out of Asttypes instead of being aliased in `parsetree.re` like we do for `mut_flag`, `rec_flag`, etc?

File: compiler/src/parsing/parsetree.re
Comment (phated) [inline compiler/src/parsing/parsetree.re:381]: See above about aliasing Asttypes

File: compiler/src/typed/typedtree.rei
Comment (phated) [inline compiler/src/typed/typedtree.rei:45]: This looks to be the alias, so if you do the Asttypes aliasing in parsetree, you don't need to change anything in this directory

**Review by phated — CHANGES_REQUESTED** [review]
This is really close! 💪 I think you can remove a bunch of your changes around moving the `Asttypes.argument_label` and that'll reduce the diff by a ton.

File: compiler/src/parsing/asttypes.re
Comment (alex-snezhko) [reply compiler/src/parsing/asttypes.re:123]: I don't actually remember why I did this. I'll revert this back into asttypes

File: stdlib/pervasives.gr
Comment (alex-snezhko) [reply stdlib/pervasives.gr:51]: Yep, at this point there aren't labels on primitives

File: stdlib/runtime/wasi.gr
Comment (alex-snezhko) [reply stdlib/runtime/wasi.gr:80]: Actually I don't think there's anything prohibiting using named arguments on foreigns, I think we just haven't gotten around to assigning argument names to all of the foreigns in the stdlib. I can do this in this PR if you'd like or alternatively in a followup PR

File: stdlib/pervasives.gr
Comment (phated) [reply stdlib/pervasives.gr:51]: #1981

File: stdlib/runtime/wasi.gr
Comment (phated) [reply stdlib/runtime/wasi.gr:80]: #1982

File: compiler/src/parsing/asttypes.re
Comment (phated) [reply compiler/src/parsing/asttypes.re]: Thanks, the diff looks great now!

**Review by phated — APPROVED** [review]

**Pull request merged**
